R/l_pkconc.R

```diff
@@ -0,0 +1,196 @@
+#'
+#' @return A list of listings, each corresponding to a unique combination of the grouping variables.
+#'
+#' @details
+#' The function performs the following steps:
+#'   - Groups the data based on the specified grouping variables.
+#'   - Formats the 0 and NA values as defined by the formatting table.
+#'   - Creates a listing for each unique combination of the grouping variables.
+#'
+#' The `formatting_vars_table` should be a data frame with the following columns:
+#'   - `var_name`: The name of the variable.
+#'   - `Label`: The label for the variable.
+#'   - `na_str`: The string to use for NA values.
+#'   - `cero_str`: The string to use for 0 values.
+#'   - `align`: The alignment for the variable (e.g., "center").
+#'   - `format_fun`: The formatting function to use ("round" or "signif").
+#'   - `digits`: The number of digits to use for numeric formatting.
+#'
+#' @examples
+#' \dontrun{
+#'   # Example usage:
+#'   listings <- l_pkconc(adpc, listgroup_vars, grouping_vars, displaying_vars)
+#' }
+#'
+#' @import dplyr formatters rlistings
+#' @export
+#' @author Gerardo Rodriguez
+l_pkconc <- function(
+  adpc = read.csv("inst/shiny/data/DummyRO_ADNCA.csv"),
+  listgroup_vars = c("PARAM", "PCSPEC", "ROUTE"),
+  grouping_vars = c("TRT01A", "USUBJID", "AVISIT"),
+  displaying_vars = c("NFRLT", "AFRLT", "AVAL"),
+  formatting_vars_table = NULL,
+  main_title = paste0("Listing of PK Concentration by Treatment Group,",
+                      "Subject and Nominal Time, PK Population"),
+  subtitle_lists = NULL,
+  footnote_table = "*: Patients excluded from the summary table and mean plots"
+) {
+
+  # If there are columns defined in the function that are not in the data, throw an error
+  missing_cols <- setdiff(c(grouping_vars, displaying_vars), colnames(adpc))
+  if (length(missing_cols) > 0) {
+    stop(paste("Missing required columns:", paste(missing_cols, collapse = ",")))
+  }
+
+  # If the title was not user defined use the listgroup_vars argument to do it
+  if (is.null(subtitle_lists)) {
+    subtitle_lists <- paste(paste0("!", listgroup_vars),
+                            paste0("$", listgroup_vars),
+                            sep = ": ",
+                            collapse = "\n")
+  }
+
+  # If the formatting table was not user defined make one standard
+  if (is.null(formatting_vars_table)) {
+    formatting_vars_table <-  data.frame(
+      var_name = c(grouping_vars, displaying_vars)
+    ) %>%
+      rowwise() %>%
+      # Create a label column
+      mutate(
+        Label = parse_annotation(adpc, paste0("!", var_name)),
+        na_str = "NA",
+        cero_str = ifelse(var_name == "AVAL", "BLQ", "0"),
+        align = "center",
+        format_fun =  ifelse(is.numeric(adpc[[var_name]]) &&
+                               !is.integer(adpc[[var_name]]), "round", NA),
+        digits = ifelse(is.numeric(adpc[[var_name]]) &&
+                          !is.integer(adpc[[var_name]]), 3, NA)
```

Comment:
suggestion: Consider `is.double`, I think it does what you intend here, unless I am missing something?
```suggestion
        format_fun =  ifelse(is.double(adpc[[var_name]]), "round", NA),
        digits = ifelse(is.double(adpc[[var_name]]), 3, NA)
```